File: typescript/src/shared/approval-rule/associate.functions.ts
import {z} from 'zod';
import {
  ApiRoot,
  ApprovalRuleDraft,
  ApprovalRuleUpdateAction,
} from '@commercetools/platform-sdk';
import {SDKError} from '../errors/sdkError';
import {CommercetoolsFuncContext} from '../../types/configuration';
import {
  readApprovalRuleById,
  readApprovalRuleByKey,
  queryApprovalRules,
  createApprovalRule as createBase,
  updateApprovalRuleById,
  updateApprovalRuleByKey,
} from './base.functions';
import {
  readApprovalRuleParameters,
  createApprovalRuleParameters,
  updateApprovalRuleParameters,
} from './parameters';

export const readApprovalRule = async (
  apiRoot: ApiRoot,
  context: CommercetoolsFuncContext,
  params: z.infer<typeof readApprovalRuleParameters>
) => {
  if (!context.customerId) {
    throw new Error(
      'Associate ID (customerId) is required for associate operations'
    );
  }
  if (!context.businessUnitKey) {
    throw new Error('Business Unit key is required for associate operations');
  }

  try {
    if (params.id) {
      return await readApprovalRuleById(
        apiRoot,
        context.projectKey,
        context.customerId,
        context.businessUnitKey,
        params.id,
        params.expand
      );
    }
    if (params.key) {
      return await readApprovalRuleByKey(
        apiRoot,
        context.projectKey,
        context.customerId,
        context.businessUnitKey,
        params.key,
        params.expand
      );
    }
    return await queryApprovalRules(
      apiRoot,
      context.projectKey,
      context.customerId,
      context.businessUnitKey,
      params.where,
      params.limit,
      params.offset,
      params.sort,
      params.expand
    );
  } catch (error: any) {
    throw new SDKError('Failed to read approval rule as associate', error);
  }
};

export const createApprovalRule = async (
  apiRoot: ApiRoot,
  context: CommercetoolsFuncContext,
  params: z.infer<typeof createApprovalRuleParameters>
) => {
  if (!context.customerId) {
    throw new Error(
      'Associate ID (customerId) is required for associate operations'
    );
  }
  if (!context.businessUnitKey) {
    throw new Error('Business Unit key is required for associate operations');
  }

  try {
    const {associateId: _a, businessUnitKey: _b, ...draft} = params;
    return await createBase(
      apiRoot,
      context.projectKey,
      context.customerId,
      context.businessUnitKey,
      draft as ApprovalRuleDraft
    );
  } catch (error: any) {
    throw new SDKError('Failed to create approval rule as associate', error);
  }
};

export const updateApprovalRule = async (
  apiRoot: ApiRoot,
  context: CommercetoolsFuncContext,
  params: z.infer<typeof updateApprovalRuleParameters>
) => {
  if (!context.customerId) {
    throw new Error(
      'Associate ID (customerId) is required for associate operations'
    );
  }
  if (!context.businessUnitKey) {
    throw new Error('Business Unit key is required for associate operations');
  }

  try {
    if (params.id) {
      return await updateApprovalRuleById(
        apiRoot,
        context.projectKey,
        context.customerId,
        context.businessUnitKey,
        params.id,
        params.version,
        params.actions as ApprovalRuleUpdateAction[]
      );
    }
    if (params.key) {
      return await updateApprovalRuleByKey(
        apiRoot,
        context.projectKey,
        context.customerId,
        context.businessUnitKey,
        params.key,
        params.version,
        params.actions as ApprovalRuleUpdateAction[]
      );
    }
    throw new Error(
      'Either id or key must be provided to update an approval rule'
    );
  } catch (error: any) {
    throw new SDKError('Failed to update approval rule as associate', error);
  }
};

File: python/commerce_mcp/tools/approval_rule/associate_functions.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any
from .schemas import ReadApprovalRuleParams, CreateApprovalRuleParams, UpdateApprovalRuleParams
from ...shared.errors import SDKError
from ...shared.transform import transform_tool_output

if TYPE_CHECKING:
    from ...api import CommercetoolsAPI
    from ...config import CTContext


def _prefix(ctx: "CTContext") -> str:
    return (
        f"/as-associate/{ctx.customer_id}"
        f"/in-business-unit/key={ctx.business_unit_key}"
    )


async def read_approval_rule(
    params: ReadApprovalRuleParams,
    api: "CommercetoolsAPI",
    ctx: "CTContext",
) -> str:
    try:
        prefix = _prefix(ctx)
        if params.id:
            query: dict[str, Any] = {}
            if params.expand:
                query["expand"] = params.expand
            result = await api.get(f"{prefix}/approval-rules/{params.id}", params=query or None)
            return transform_tool_output(result)
        if params.key:
            query = {}
            if params.expand:
                query["expand"] = params.expand
            result = await api.get(f"{prefix}/approval-rules/key={params.key}", params=query or None)
            return transform_tool_output(result)
        query = {"limit": params.limit or 10}
        if params.where:
            query["where"] = params.where
        if params.offset is not None:
            query["offset"] = params.offset
        if params.sort:
            query["sort"] = params.sort
        if params.expand:
            query["expand"] = params.expand
        result = await api.get(f"{prefix}/approval-rules", params=query)
        return transform_tool_output(result)
    except Exception as e:
        raise SDKError("read approval rule as associate", e)


async def create_approval_rule(
    params: CreateApprovalRuleParams,
    api: "CommercetoolsAPI",
    ctx: "CTContext",
) -> str:
    try:
        prefix = _prefix(ctx)
        body: dict[str, Any] = {
            "name": params.name,
            "predicate": params.predicate,
            "approvers": params.approvers,
            "requesters": params.requesters,
            "status": params.status,
        }
        if params.key:
            body["key"] = params.key
        if params.description:
            body["description"] = params.description
        result = await api.post(f"{prefix}/approval-rules", body=body)
        return transform_tool_output(result)
    except Exception as e:
        raise SDKError("create approval rule as associate", e)


async def update_approval_rule(
    params: UpdateApprovalRuleParams,
    api: "CommercetoolsAPI",
    ctx: "CTContext",
) -> str:
    try:
        prefix = _prefix(ctx)
        body: dict[str, Any] = {"version": params.version, "actions": params.actions}
        if params.id:
            result = await api.post(f"{prefix}/approval-rules/{params.id}", body=body)
            return transform_tool_output(result)
        if params.key:
            result = await api.post(f"{prefix}/approval-rules/key={params.key}", body=body)
            return transform_tool_output(result)
        raise ValueError("Either id or key must be provided to update an approval rule")
    except ValueError:
        raise
    except Exception as e:
        raise SDKError("update approval rule as associate", e)
